src/coreclr/System.Private.CoreLib/src/System/Threading/Mutex.CoreCLR.Unix.cs

```diff
@@ -0,0 +1,135 @@
+        {
+            SafeWaitHandle mutexHandle = CreateMutexCore(initiallyOwned, name, out int errorCode, out string? errorDetails);
+            if (mutexHandle.IsInvalid)
+            {
+                mutexHandle.SetHandleAsInvalid();
+                if (errorCode == Interop.Errors.ERROR_FILENAME_EXCED_RANGE)
+                    // On Unix, length validation is done by CoreCLR's PAL after converting to utf-8
+                    throw new ArgumentException(SR.Argument_WaitHandleNameTooLong, nameof(name));
+                if (errorCode == Interop.Errors.ERROR_INVALID_HANDLE)
+                    throw new WaitHandleCannotBeOpenedException(SR.Format(SR.Threading_WaitHandleCannotBeOpenedException_InvalidHandle, name));
+
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, name, errorDetails);
+            }
+
+            createdNew = errorCode != Interop.Errors.ERROR_ALREADY_EXISTS;
+            SafeWaitHandle = mutexHandle;
+        }
+
+        private static OpenExistingResult OpenExistingWorker(string name, out Mutex? result)
+        {
+            ArgumentException.ThrowIfNullOrEmpty(name);
+
+            result = null;
+            // To allow users to view & edit the ACL's, call OpenMutex
+            // with parameters to allow us to view & edit the ACL.  This will
+            // fail if we don't have permission to view or edit the ACL's.
+            // If that happens, ask for less permissions.
+            SafeWaitHandle myHandle = OpenMutexCore(name, out int errorCode, out string? errorDetails);
+
+            if (myHandle.IsInvalid)
+            {
+                myHandle.Dispose();
+
+                if (errorCode == Interop.Errors.ERROR_FILENAME_EXCED_RANGE)
+                {
+                    // On Unix, length validation is done by CoreCLR's PAL after converting to utf-8
+                    throw new ArgumentException(SR.Argument_WaitHandleNameTooLong, nameof(name));
+                }
+                if (Interop.Errors.ERROR_FILE_NOT_FOUND == errorCode || Interop.Errors.ERROR_INVALID_NAME == errorCode)
+                    return OpenExistingResult.NameNotFound;
+                if (Interop.Errors.ERROR_PATH_NOT_FOUND == errorCode)
+                    return OpenExistingResult.PathNotFound;
+                if (Interop.Errors.ERROR_INVALID_HANDLE == errorCode)
+                    return OpenExistingResult.NameInvalid;
+
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, name, errorDetails);
+            }
+
+            result = new Mutex(myHandle);
+            return OpenExistingResult.Success;
+        }
+
+        // Note: To call ReleaseMutex, you must have an ACL granting you
+        // MUTEX_MODIFY_STATE rights (0x0001). The other interesting value
+        // in a Mutex's ACL is MUTEX_ALL_ACCESS (0x1F0001).
+        public void ReleaseMutex()
+        {
+            if (!Interop.Kernel32.ReleaseMutex(SafeWaitHandle))
```

Comment:
is error code ever worth getting / exposting from releasemutex? since you're preserving errors in other places.

Reply:
Natural errors aren't expected from `ReleaseMutex`. There can be errors from acquiring the lock (waiting on a mutex) and that's not covered, but most of the issues are likely to be found when creating/opening a mutex. Covering waits is more involved.